src/repr/src/row/collection.rs

```diff
@@ -225,32 +267,143 @@ impl RustType<ProtoEncodedRowMetadata> for EncodedRowMetadata {
     }
 }
 
-/// Provides a sorted view of a [`RowCollection`].
+/// Used to store a [`RowCollection`] in a [`BinaryHeap`] without creating any references.
+/// Each row in the row collection is successively "inserted" into the heap by simply updating
+/// `row_index`. Once all entries are exhausted, the `HeapRowCollection` is removed from the heap.
+///
+/// Implements a custom [`Ord`] to make the heap a _min-heap_.
 #[derive(Debug, Clone)]
-pub struct SortedRowCollection {
-    /// The inner [`RowCollection`].
-    collection: RowCollection,
-    /// Indexes into the inner collection that represent the sorted order.
-    sorted_view: Arc<[usize]>,
+pub struct HeapRowCollection {
+    /// A sorted [`RowCollection`].
+    row_collection: RowCollection,
+    /// The `row_collection` index that is considered active in the heap.
+    row_index: usize,
+    /// The number of diffs left to process of the current row.
+    diffs_left: usize,
 }
 
-impl SortedRowCollection {
-    pub fn get(&self, idx: usize) -> Option<(&RowRef, &EncodedRowMetadata)> {
-        self.sorted_view
-            .get(idx)
-            .and_then(|inner_idx| self.collection.get(*inner_idx))
+impl HeapRowCollection {
+    /// Helper function to retrieve the currently active [`RowRef`].
+    fn get_row(&self) -> Option<&RowRef> {
+        self.row_collection.get_row(self.row_index)
+    }
+}
+
+impl PartialEq for HeapRowCollection {
+    fn eq(&self, other: &Self) -> bool {
+        self.get_row().eq(&other.get_row())
+    }
+}
+// Cannot derive [`Eq`] as [`RowCollection`] does not implement it.
+impl Eq for HeapRowCollection {}
+
+impl PartialOrd for HeapRowCollection {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for HeapRowCollection {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // The comparison is done the other way to ensure a _min-heap_ when inserting into a
+        // [`BinaryHeap`], which is a _max-heap_ by default.
+        other.get_row().cmp(&self.get_row())
```

Comment:
This uses the row sort order, which isn't the same as the order specified by the `ORDER BY` clause. We need to use the row set finishing here, and compare the rows based on that order.

Reply:
Thanks, fixed.